python/ray/data/_internal/util.py

```diff
@@ -143,13 +150,51 @@ def _autodetect_parallelism(
         for the detected parallelism (only if -1 was specified), and the estimated
         inmemory size of the dataset.
     """
-    min_safe_parallelism = 1
-    max_reasonable_parallelism = sys.maxsize
+
     if mem_size is None and datasource_or_legacy_reader:
         mem_size = datasource_or_legacy_reader.estimate_inmemory_data_size()
+
+    target_min_block_size = ctx.target_min_block_size
+
+    if target_max_block_size < target_min_block_size:
+        logger.warning(
+            f"Provided `target_max_block_size` ({target_max_block_size / MiB:.1f}MiB) "
+            f"< target_min_block_size` ({target_min_block_size / MiB:.1f}); "
+            f"overriding `target_min_block_size` with `target_max_block_size`"
+        )
+
+        target_min_block_size = target_max_block_size
+
+    def _compose_memory_estimation_context() -> str:
+        estimated_size_in_memory_str = (
+            f"{mem_size / MiB:.2f}MiB" if mem_size is not None else "NaN"
+        )
+
+        return (
+            f"Estimated data size {estimated_size_in_memory_str}; "
+            f"Target min/max block sizes are "
+            f"{target_min_block_size / MiB:.1f}MiB / "
+            f"{target_max_block_size / MiB:.1f}MiB"
+        )
+
+    # Set fallback values for parallelism boundaries
+    min_safe_parallelism = 1
+    max_reasonable_parallelism = sys.maxsize
+
+    # Evaluate safe parallelism bounds based on estimated data size
+    #   - Min (safe) parallelism is determined such that produced blocks
+    #   do not exceed `target_max_block_size`
+    #   - Max (reasonable) parallelism is determined such that produced blocks
+    #   are no smaller than `DataContext.target_min_block_size`
     if mem_size is not None and not np.isnan(mem_size):
         min_safe_parallelism = max(1, int(mem_size / target_max_block_size))
-        max_reasonable_parallelism = max(1, int(mem_size / ctx.target_min_block_size))
+        max_reasonable_parallelism = max(1, int(mem_size / target_min_block_size))
+
+        estimation_context = _compose_memory_estimation_context()
+
+        assert (
+            min_safe_parallelism <= max_reasonable_parallelism
+        ), f"Parallelism boundaries have to overlap: {estimation_context}"
```

Comment:
nit, use an if statement to avoid generating the error message string when not needed.